lib/AST/Decl.cpp

```diff
@@ -7488,17 +7491,86 @@ bool AbstractFunctionDecl::isDistributedActorSystemRemoteCall(bool isVoidReturn)
     return false;
   }
 
-  // FIXME(distributed): check the exact generic requirements
+  // Generic parameters
+  // GenericTypeParamDecl *ActParam = genericParams->getParams()[0];
+  // GenericTypeParamDecl *ErrParam = genericParams->getParams()[1];
+  GenericTypeParamDecl *ResParam =
+      isVoidReturn ? nullptr : genericParams->getParams().back();
+
+  size_t expectedRequirementsNum = 3;
+  if (!isVoidReturn) {
+    // TODO(distributed): support alternative SerializationRequirements here
```

Comment:
Here is an alternative way to check this:

1. Lookup declaring type - should be an implementation of `DistributedActorSystem`
2. Find a concrete serialization requirement associated with it e.g. `Codable`
3. Find a concrete `InvocationDecoder` type
3. Extract either a single protocol type or a set of protocols (if requirement is a ProtocolCompositionType)
4. Structural checks:
    - async + throws
    - Has 3 generic parameters
    - Has required number of parameters,
    - labels match (!)
    - result type is represented by a generic parameter or `Void`
5. Check that parameter at index `0` is a generic parameter and it conforms to `DistributedActor`
6. Check that parameter at index `1` is a known type - `RemoteCallTarget` (you'd have to make that type known to the compiler)
7. Check that parameter at index `2` is equal to concrete type of `InvocationDecoder` and parameter is `inout`
8. Check that parameter at index `3` is a metatype of a generic parameter or `Void` and if it's former that it "covers" all of the protocols from step 3 (you can get a set of requirements associated with a given generic parameter via `GenericSignature::RequiredProtocols getRequiredProtocols(Type type) const;`).
9. Check that parameter at index `4` is a metatype of generic parameter and generic parameter has a requirement for `Error` (same `getRequiredProtocols` call).

Reply:
You could either check the label list of equivalence or do per-parameter label checks.

Reply:
Thanks -- getting back to this now :)

Reply:
Thanks for the hints this makes it much cleaner; Only  halfway through so far since i implemented dealing with the serialization req a bit more. Shaping up well tho

Reply:
Done and more cleanups along the way.